packages/http-proxy-agent/src/index.ts

```diff
@@ -124,7 +141,9 @@ export class HttpProxyAgent<Uri extends string> extends Agent {
 		let first: string;
 		let endOfHeaders: number;
 		debug('Regenerating stored HTTP header string for request');
-		req._implicitHeader();
+		if (!req._header) {
+			req._implicitHeader();
+		}
```

Comment:
I had to add this conditional because now that we are setting the headers before `addRequest()`, it causes `req._header` to be non-null here, which will cause `req._implicitHeader()` to [throw an exception](https://github.com/nodejs/node/blob/f36461dc6a0a334cbe8e427788377826232f3e4f/lib/_http_client.js#L395-L398).

Alternatively, I could do just allow `connect()` to rewrite the request URL and set headers, even if it's already happened in `addRequest()`:

```diff
diff --git a/packages/http-proxy-agent/src/index.ts b/packages/http-proxy-agent/src/index.ts
index 16d5b51..807e4c5 100644
--- a/packages/http-proxy-agent/src/index.ts
+++ b/packages/http-proxy-agent/src/index.ts
@@ -29,7 +29,6 @@ interface HttpProxyAgentClientRequest extends http.ClientRequest {
 	}[];
 	_header?: string | null;
 	_implicitHeader(): void;
-	_updatedPropsForProxyAgent?: boolean;
 }
 
 function isHTTPS(protocol?: string | null): boolean {
@@ -104,8 +103,6 @@ export class HttpProxyAgent<Uri extends string> extends Agent {
 				this.keepAlive ? 'Keep-Alive' : 'close'
 			);
 		}
-
-		req._updatedPropsForProxyAgent = true;
 	}
 
 	addRequest(req: HttpProxyAgentClientRequest, opts: AgentConnectOpts): void {
@@ -118,12 +115,10 @@ export class HttpProxyAgent<Uri extends string> extends Agent {
 		req: HttpProxyAgentClientRequest,
 		opts: AgentConnectOpts
 	): Promise<net.Socket> {
-		if (!req._updatedPropsForProxyAgent) {
-			// Maintain backwards compatibility with ProxyAgent and
-			// PacProxyAgent, which both rely on HttpProxyAgent to set request
-			// props inside the connect() function.
-			this.setRequestProps(req, opts);
-		}
+		// Maintain backwards compatibility with ProxyAgent and
+		// PacProxyAgent, which both rely on HttpProxyAgent to set request
+		// props inside the connect() function.
+		this.setRequestProps(req, opts);
 
 		// Create a socket connection to the proxy server.
 		let socket: net.Socket;
@@ -141,9 +136,7 @@ export class HttpProxyAgent<Uri extends string> extends Agent {
 		let first: string;
 		let endOfHeaders: number;
 		debug('Regenerating stored HTTP header string for request');
-		if (!req._header) {
-			req._implicitHeader();
-		}
+		req._implicitHeader();
 		if (req.outputData && req.outputData.length > 0) {
 			// Node >= 12
 			debug(
```

What do you think?